server/src/main/java/org/elasticsearch/ingest/PipelineUtils.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.ingest;
+
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.cluster.metadata.IndexAbstraction;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.cluster.metadata.IndexTemplateMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.MetadataIndexTemplateService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexSettings;
+
+import java.util.List;
+import java.util.Objects;
+
+public class PipelineUtils {
+
+    public static final String NOOP_PIPELINE_NAME = "_none";
```

Comment:
I'm not sold on a utils class in general, but regardless of whether we go that way or not, I'm not in favor of it owning this constant -- I think you should put it back on `IngestService`.

Reply:
I neither, but this `IngestService` class is a bit bloated with a lot of logic, and doing this method split (into resolve + calculate methods) would increase the needed cognitive load. I'll move back to `IngestService` the logic.

As you know, I'm still in the process of learning elastic's mores.